src/realm/mixed.cpp

```diff
@@ -85,88 +85,207 @@ inline int compare_float(Float a_raw, Float b_raw)
+                case type_Decimal:
+                    return -_impl::compare_decimal_to_double(b.decimal_val, double_val);
+                default:
+                    break;
+            }
             break;
         case type_Timestamp:
-            if (get<Timestamp>() > b.get<Timestamp>())
-                return 1;
-            else if (get<Timestamp>() < b.get<Timestamp>())
-                return -1;
+            if (b.get_type() == type_Timestamp) {
+                return _impl::compare_generic(date_val, b.date_val);
+            }
             break;
-        case type_ObjectId: {
-            auto l = get<ObjectId>();
-            auto r = b.get<ObjectId>();
-            if (l > r)
-                return 1;
-            else if (l < r)
-                return -1;
+        case type_ObjectId:
+            if (b.get_type() == type_ObjectId) {
+                return _impl::compare_generic(id_val, b.id_val);
+            }
             break;
-        }
-        case type_Decimal: {
-            auto l = get<Decimal128>();
-            auto r = b.get<Decimal128>();
-            if (l > r)
-                return 1;
-            else if (l < r)
-                return -1;
+        case type_Decimal:
+            switch (b.get_type()) {
+                case type_Int:
+                    return _impl::compare_decimals(decimal_val, Decimal128(b.int_val));
+                case type_Bool:
+                    return _impl::compare_decimals(decimal_val, Decimal128(b.bool_val ? 1 : 0));
+                case type_Float:
+                    return _impl::compare_decimal_to_double(decimal_val, double(b.float_val));
+                case type_Double:
+                    return _impl::compare_decimal_to_double(decimal_val, b.double_val);
+                case type_Decimal:
+                    return _impl::compare_decimals(decimal_val, b.decimal_val);
+                default:
+                    break;
+            }
             break;
-        }
         case type_Link:
-            if (get<ObjKey>() > b.get<ObjKey>())
-                return 1;
-            else if (get<ObjKey>() < b.get<ObjKey>())
-                return -1;
+            if (b.get_type() == type_Link) {
+                return _impl::compare_generic(int_val, b.int_val);
+            }
             break;
         case type_TypedLink:
-            if (get<ObjLink>() > b.get<ObjLink>())
-                return 1;
-            else if (get<ObjLink>() < b.get<ObjLink>())
-                return -1;
+            if (b.get_type() == type_TypedLink) {
+                return _impl::compare_generic(link_val, b.link_val);
+            }
             break;
         default:
             REALM_ASSERT_RELEASE(false && "Compare not supported for this column type");
             break;
     }
 
-    return 0;
+    // Comparing types ass a fallback option makes it possible to make a sort of a list of Mixed
```

Comment:
It's unclear to me what a types "ass" is... Can you please elaborate? Perhaps with an ASCII art illustration?

Reply:
Fixed